### DISCUSSION TITLE
Should there be an isOnline addon?

### DISCUSSION THREAD

**Discussion by jeffalo** [ScratchAddons/ScratchAddons#342] [Discussion]
so i remember back when isOnline was at its peak and it was super useful to tell if someone was online.

i was wondering, maybe there could be a super simple iO implementation (probably not with the recommended tag and with a new name) which wouldn't have any of the friends list features.

basically, ping server every few minutes with username (or maybe user id), and on the server, if someone has not pinged in x minutes, then they are offline.

then when visiting a profile, fetch an api from the server and decide whether they are online, offline, or not using the extension. 

isonline had some other features that let you know if your friends were online, but something like that could cause privacy concerns blah blah. it also had a ghost mode, which would also be fairly simple to implement in this simple iO.

i think even i know how to make a simple iO implementation and if you want i can make the PR to add it.

_there might also have to be a privacy policy and stuff like that 😬_

**Comment by jeffalo** [comment]
i made a demo server on repl.it

https://io-meme.jeffalobob.repl.co/ will display how many users are online
https://io-meme.jeffalobob.repl.co/online?user=username will make a user online for 10 seconds
https://io-meme.jeffalobob.repl.co/isonline?user=username will check if they're online

**Comment by jeffalo** [comment]
and the code https://repl.it/@jeffalobob/io-meme

**Comment by jemunm378** [comment]
I think it’s a great idea.

**Comment by tb148** [comment]
This is AWESOME! However we should also have a "non-SA user" status so people don't get confused why online users have an "offline" status.
This might be done by setting an `isUser` status the first time they ping. Defaults to `false`, sets to `true` on their first ping, and reset to `false` when they disable/uninstall.

**Comment by retronbv** [comment]
Yas

**Comment by jeffalo** [comment]
we will also have to handle verification if the user pinging is really them. any ideas on how to implement this?

**Comment by jeffalo** [comment]
@WorldLanguages ^^

**Comment by jemunm378** [comment]
Maybe the users can verify using their scratch account. Maybe something similar to this?: https://github.com/ScratchVerifier/ScratchVerifier

**Comment by WorldLanguages** [comment]
Keep in mind, we won't be adding any online indicator/notifications for now, while we focus in other parts of the extension.

**Comment by jeffalo** [comment]
i can focus though

**Comment by WorldLanguages** [comment]
I need to focus on whether your focus is appropriate :P

**Comment by jeffalo** [comment]
i was thinking of something that may be better, instead of displaying offline users, only online users are displayed. 

this makes it a bit simpler to code and allows focus or something idk just let me make it

**Comment by Explosion-Scratch** [comment]
This could be very neatly incorporated into the multiplayer cloud games section of griffpatch's messaging extension. It would be nice though if it only showed who was online, not who was offline and also maybe the option to be offline even when online. (Perhaps a prompt could solve this.)

**Comment by williamhorning** [comment]
I was thinking about how to do this, and I have a solution.
if the user turns on the addon, show other users online / offline
get data from a scratch project that proxies requests to a server
add an option to share online / offline
this option would also go through the proxy project

**Reply by Riptide-Scratch** [reply]
IDK if they are able to get server hosting...

**Reply by williamhorning** [reply]
> IDK if they are able to get server hosting...

I have a server already set up with this code

**Comment by MESYETI** [comment]
> so i remember back when isOnline was at its peak and it was super useful to tell if someone was online.
> 
> i was wondering, maybe there could be a super simple iO implementation (probably not with the recommended tag and with a new name) which wouldn't have any of the friends list features.
> 
> basically, ping server every few minutes with username (or maybe user id), and on the server, if someone has not pinged in x minutes, then they are offline.
> 
> then when visiting a profile, fetch an api from the server and decide whether they are online, offline, or not using the extension.
> 
> isonline had some other features that let you know if your friends were online, but something like that could cause privacy concerns blah blah. it also had a ghost mode, which would also be fairly simple to implement in this simple iO.
> 
> i think even i know how to make a simple iO implementation and if you want i can make the PR to add it.
> 
> _there might also have to be a privacy policy and stuff like that grimacing_

no its an invasion of privacy id rather agree with facbook privacy policy!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!1111111111

**Reply by MESYETI** [reply]
good argument mesyeti 10/10

**Comment by FunctionalMetatable** [comment]
How about using fosshost like turbowarp uses?

**Reply by FunctionalMetatable** [reply]
as in hosting the server through fosshost

**Comment by FunctionalMetatable** [comment]
An idea of an implementation for an auth system:
1. When the addon is enabled, open a new popup that goes to OAuth2.
2. When the user finishes auth, possibly redirect to a server and that server generates a secure token.
3. With that secure token, it will redirect the user to https://scratch.mit.edu/is-online/verify. The query parameter "auth" would be the token in this url.
4. The isOnline addon would get that token and put it in scratch's cookie.
5. Every 1 minute the addon will ping a server with the isonline token in the headers.
6. To log out of isOnline, the addon will send the token to the server in its headers. The server will take the token out of its database
